src/Microsoft.FeatureManagement/VariantServiceProvider.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// Used to get different implementations of TService depending on the assigned variant from a specific variant feature flag.
+    /// </summary>
+    internal class VariantServiceProvider<TService> : IVariantServiceProvider<TService> where TService : class
+    {
+        private readonly IEnumerable<TService> _services;
+        private readonly IVariantFeatureManager _featureManager;
+        private readonly string _variantFeatureName;
+
+        /// <summary>
+        /// Creates a variant service provider.
+        /// </summary>
+        /// <param name="services">Implementation variants of TService.</param>
+        /// <param name="featureManager">Feature manager to get the assigned variant of the variant feature flag.</param>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="services"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">Thrown if <paramref name="featureManager"/> is null.</exception>
+        public VariantServiceProvider(IEnumerable<TService> services, IVariantFeatureManager featureManager)
+        {
+            _services = services ?? throw new ArgumentNullException(nameof(services));
+            _featureManager = featureManager ?? throw new ArgumentNullException(nameof(featureManager));
+        }
+
+        /// <summary>
+        /// The variant feature flag used to assign variants.
+        /// </summary>
+        public string VariantFeatureName
+        {
+            get => _variantFeatureName;
+
+            init
+            {
+                _variantFeatureName = value ?? throw new ArgumentNullException(nameof(value));
+            }
+        }
+
+        /// <summary>
+        /// Gets implementation of TService according to the assigned variant from the feature flag.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token to cancel the operation.</param>
+        /// <returns>An implementation matched with the assigned variant. If there is no matched implementation, it will return null.</returns>
+        public async ValueTask<TService> GetServiceAsync(CancellationToken cancellationToken)
+        {
+            Debug.Assert(_variantFeatureName != null);
+
+            Variant variant = await _featureManager.GetVariantAsync(_variantFeatureName, cancellationToken);
+
+            TService implementation = null;
+
+            if (variant != null)
+            {
+                implementation = _services.FirstOrDefault(service => 
```

Comment:
We might want to do this work upfront- if we added an "Init()" function, we could call it in our BuilderExtension and have it place the services into a Map<string, TService> where the string is either the alias or the name. 

This way we don't have to do reflection on every request/time GetServiceAsync is called.

What do you think?

Reply:
> if we added an "Init()" function, we could call it in our BuilderExtension and have it place the services into a Map<string, TService> where the string is either the alias or the name.

This sounds not feasible for me. Could you provide more details about how this can be achieved?

If we want to have a Map<string, TService>, the value of each key is a TService instance. This could not be done during the service registration stage. In our BuilderExtension, the registered service are actually `ServiceDescriptor` in `IServiceCollection` and they are still not instantiated.

Reply:
Added in 7120abd97a0ee8ed44157d174198b0aeabc926d6  @rossgrambo